File: MAStaff-Bukkit/src/main/java/es/angelillo15/mast/bukkit/listener/clickListeners/OnItemClick.java
package es.angelillo15.mast.bukkit.listener.clickListeners;

import com.google.inject.Inject;
import es.angelillo15.mast.api.IStaffPlayer;
import es.angelillo15.mast.api.Permissions;
import es.angelillo15.mast.api.items.IExecutableItem;
import es.angelillo15.mast.api.items.IExecutableLocationItem;
import es.angelillo15.mast.api.items.StaffItem;
import es.angelillo15.mast.api.managers.StaffManager;
import es.angelillo15.mast.bukkit.MAStaff;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.player.PlayerInteractEvent;

public class OnItemClick implements Listener {
    @Inject
    private StaffManager staffManager;

    @EventHandler
    public void onItemClick(PlayerInteractEvent event) {

        Player player = event.getPlayer();

        if (!player.hasPermission(Permissions.STAFF.getPermission())) {
            return;
        }

        IStaffPlayer staffPlayer = staffManager.getStaffPlayer(player);

        if (!staffPlayer.isStaffMode()) {
            return;
        }

        if (event.getItem() == null) {

            if (!player.hasPermission(Permissions.STAFF_BUILD.getPermission())) {
                event.setCancelled(true);
            }
            return;
        }

        StaffItem item = staffPlayer.getItems().get(player.getItemInHand().getItemMeta().getDisplayName());

        if (item instanceof IExecutableItem) {
            MAStaff.getPlugin().getPLogger().debug("Executing item " +
                    item.getItem().getItemMeta().getDisplayName() +
                    " for player " + player.getName()
            );
            ((IExecutableItem) item).click(player);
        }

        if (item instanceof IExecutableLocationItem) {
            if (event.getClickedBlock() == null) return;

            MAStaff.getPlugin().getPLogger().debug("Executing item " +
                    item.getItem().getItemMeta().getDisplayName() +
                    " for player " + player.getName()
            );

            ((IExecutableLocationItem) item).click(player, event.getClickedBlock().getLocation());
        }

        event.setCancelled(true);
    }
}

File: MAStaff-Bukkit/src/main/java/es/angelillo15/mast/bukkit/listener/clickListeners/OnItemClick.kt
package es.angelillo15.mast.bukkit.listener.clickListeners

import com.google.inject.Inject
import es.angelillo15.mast.api.ILogger
import es.angelillo15.mast.api.Permissions
import es.angelillo15.mast.api.items.IExecutableItem
import es.angelillo15.mast.api.items.IExecutableLocationItem
import es.angelillo15.mast.api.managers.StaffManager
import es.angelillo15.mast.api.nms.VersionSupport
import es.angelillo15.mast.api.thread.execute
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.player.PlayerInteractEvent

class OnItemClick : Listener {
  @Inject
  private lateinit var staffManager: StaffManager

  @Inject
  private lateinit var versionSupport: VersionSupport

  @Inject
  private lateinit var logger: ILogger

  @EventHandler
  fun onItemClick(event: PlayerInteractEvent) {
    val start = System.currentTimeMillis()
    val player = event.player

    if (!staffManager.isStaffPlayer(player)) {
      return
    }

    val staffPlayer = staffManager.getStaffPlayer(player) ?: return

    if (!staffPlayer.isStaffMode()) {
      return
    }

    if (event.item == null) {
      if (!player.hasPermission(Permissions.STAFF_BUILD.permission)) {
        event.setCancelled(true)
      }
      return
    }

    val tag = versionSupport.getTag(event.item!!, "mast-staff-item")

    if (tag == null) {
      event.setCancelled(true)
      return
    }

    logger.debug("Item " + tag + " clicked by " + player.name)

    val item = staffPlayer.getItems()[tag]

    if (item == null) {
      event.setCancelled(true)
      return
    }

    if (item is IExecutableItem) {
      (item as IExecutableItem).click(player)
    }

    if (item is IExecutableLocationItem) {
      if (event.clickedBlock == null) return
      val block = event.clickedBlock!!

      execute {
        (item as IExecutableLocationItem).click(player, block.location)
      }
    }
    logger.debug("Item " + item.getName() + " took " + (System.currentTimeMillis() - start) + "ms to execute")
    event.setCancelled(true)
  }
}
